### ISSUE TITLE
MemoryStream.map(fn).remember().map() reruns fn despite remember. 

### ISSUE THREAD

**Issue by Steelfish** [staltz/xstream#108]
I've worked out the issue in a webpackbin: http://www.webpackbin.com/NkaAtqAY-

For as far as I can tell, calling remember on a mapped memorystream and calling map again does not cause the output stream to remember the transformed value, resulting in the transform function being run for every listener.
Calling just .map(fn).remember() on a memorystream does however remember the transformed value.
When using remember on a mapped stream, the transformed value is remembered as expected. 

I believe this is related to #98 and #93. 

After some digging I believe it's due to the composition operator for Map operators. f(g(x)) works fine for regular streams but causes g to be run twice for memorystreams whereas I'd expected it to be run only once and remembered. In my opinion the map operator should either only return streams, allowing remember() to return a remembered g(x) or return a Map(f, Remember(g(x))) for MemoryStreams when dealing with composition.

**Comment by xtianjohns** [comment]
Memoizing the MapOperator project function solves this problem for me.

```ts
// MapOperator constructor
  constructor(project: (t: T) => R, ins: Stream<T>) {
    // ...
    if ( ins instanceof MemoryStream ) {
        // We have a memory stream.
        this.project = memo( project );
    } else {
        this.project = project;
    }
  }
// ...
```
Is this a desirable architectural approach? If so, is there a great place to park `memo`, perhaps on the MapOperator itself?

**Comment by staltz** [comment]
Fixed in v7.0
